# FIT5197 2024 S1 Final Assessment

**SPECIAL NOTE:** Please refer to the [assessment page](https://learning.monash.edu/mod/assign/view.php?id=2017255) for rules, general guidelines and marking rubrics of the assessment (the marking rubric for the kaggle competition part will be released near the deadline in the same page). Failure to comply with the provided information will result in a deduction of mark (e.g., late penalties) or breach of academic integrity.

**YOUR NAME**: Quoc Khoa Tran

**STUDENT ID**: 34124888

**KAGGLE NAME/ID** (See part 1, Question 5 or part 2, there are penalties if you don't enter it here!!!): Quoc Khoa Tran

Please also enter your details in this [google form](https://forms.gle/isxqfrVnV7ddAj8y8).

# Part 1 Regression (50 Marks)

A few thousand people were questioned in a [life and wellbeing survey](https://www.get-happier.com/) to build a model to predict happiness of an individual. You need to build regression models to optimally predict the variable in the survey dataset called 'happiness' based on any, or all, of the other survey question responses. 

You have been provided with two datasets, ```regression_train.csv``` and ```regression_test.csv```. Using these datasets, you hope to build a model that can predict happiness level using the other variables. ```regression_train.csv``` comes with the ground-truth target label (i.e. happiness level) whereas `regression_test.csv` comes with independent variables (input information) only.

On the order of around 70 survey questions have been converted into predictor variables that can be used to predict happiness. We do not list all the predictor names here, but their names given in the data header can clearly be linked to the survey questions. e.g., the predictor variable 'iDontFeelParticularlyPleasedWithTheWayIAm' corresponds to the survey question 'I don’t feel particularly pleased with the way I am.'

**PLEASE NOTE THAT THE USE OF LIBRARIES ARE PROHIBITED IN THESE QUESTIONS UNLESS STATED OTHERWISE, ANSWERS USING LIBRARIES WILL RECEIVE 0 MARKS**

## Question 1 (NO LIBRARIES ALLOWED) (4 Mark)
Please load the ```regression_train.csv``` and fit a [$\textbf{multiple linear regression model}$](https://en.wikipedia.org/wiki/Linear_regression) with 'happiness' being the target variable. According to the summary table, which predictors do you think are possibly associated with the target variable (use the significance level of 0.01), and which are the **Top 5** strongest predictors? Please write an R script to automatically fetch and print this information.

**NOTE**: Manually doing the above tasks will result in 0 marks.

In [10]:
# ANSWER BLOCK

# Load the data manually without using libraries
train_df <- read.csv("regression_train.csv")

# Fit a multiple linear regression model
model <- lm(happiness ~ ., data = train_df)

# Summarize the model to get the significance of predictors
model_summary <- summary(model)

# Extract the coefficients table
coefficients_table <- model_summary$coefficients

# Identify significant predictors at the 0.01 level
significant_predictors <- coefficients_table[coefficients_table[, "Pr(>|t|)"] < 0.01, ]

# Identify top 5 strongest predictors based on absolute t-value
top_5_predictors <- head(significant_predictors[order(abs(significant_predictors[, "t value"]), decreasing = TRUE), ], 5)

# Print the results
print("Significant predictors at 0.01 level:")
print(significant_predictors)

print("Top 5 strongest predictors:")
print(top_5_predictors)

[1] "Significant predictors at 0.01 level:"
                                                       Estimate Std. Error
(Intercept)                                          -22.138272  3.5197268
income10k - 15k                                        6.268117  1.2386952
income120k - 150k                                     11.373025  1.2340393
income150k - 200k                                     11.718396  1.2434310
income15k - 20k                                       15.447908  1.2809264
income200k above                                      20.109900  1.3788320
income20k - 50k                                       20.839607  1.5825100
income50k - 80k                                       28.546980  1.7382901
income80k - 120k                                      37.315988  1.6548496
whatIsYourHeightExpressItAsANumberInMetresM175 - 180   6.104323  2.2061513
whatIsYourHeightExpressItAsANumberInMetresM180 - 185   7.286107  2.3791630
alwaysStressed                                        -1

## Question 2 (2 Mark)
[**R squared**](https://en.wikipedia.org/wiki/Coefficient_of_determination) from the summary table reflects that the full model doesn't fit the training dataset well; thus, you try to quantify the error between the values of the ground-truth and those of the model prediction. You want to write a function to predict 'happiness' with the given dataset and calculate the [root mean squared error (rMSE)](https://en.wikipedia.org/wiki/Root-mean-square_deviation) between the model predictions and the ground truths. Please test this function on the full model and the training dataset.

In [11]:
# ANSWER BLOCK
# Function to calculate RMSE
calculate_rmse <- function(actual, predicted) {
  sqrt(mean((actual - predicted)^2))
}

# Predict happiness on the training data
predictions <- predict(model, newdata = train_df)

# Calculate RMSE
rmse_full_model <- calculate_rmse(train_df$happiness, predictions)

# Print RMSE
print(rmse_full_model)

[1] 6.672557


## Question 3 (2 Marks)
You find the full model complicated and try to reduce the complexity by performing [bidirectional stepwise regression](https://en.wikipedia.org/wiki/Stepwise_regression) with [BIC](https://en.wikipedia.org/wiki/Bayesian_information_criterion).

Calculate the **rMSE** of this new model with the function that you implemented previously. Is there anything you find unusual? Explain your findings in 100 words.

In [12]:
# ANSWER BLOCK
# Perform bidirectional stepwise regression using BIC
stepwise_model <- step(model, direction = "both", k = log(nrow(train_df)))

# Predict happiness on the training data using the stepwise model
stepwise_predictions <- predict(stepwise_model, newdata = train_df)

# Calculate RMSE for the stepwise model
rmse_stepwise_model <- calculate_rmse(train_df$happiness, stepwise_predictions)

# Print RMSE for the stepwise model
print(paste("Root Mean Squared Error (rMSE) of Stepwise Model:", rmse_stepwise_model))

Start:  AIC=2326.81
happiness ~ gender + income + whatIsYourHeightExpressItAsANumberInMetresM + 
    doYouFeelASenseOfPurposeAndMeaningInYourLife104 + howDoYouReconcileSpiritualBeliefsWithScientificOrRationalThinki + 
    howOftenDoYouFeelSociallyConnectedWithYourPeersAndFriends + 
    doYouHaveASupportSystemOfFriendsAndFamilyToTurnToWhenNeeded + 
    howOftenDoYouParticipateInSocialActivitiesIncludingClubsSportsV + 
    doYouFeelComfortableEngagingInConversationsWithPeopleFromDiffer + 
    doYouFeelASenseOfPurposeAndMeaningInYourLife105 + alwaysAnxious + 
    alwaysStressed + alwaysAccountableAndResponsibleForYourActions + 
    alwaysCalm + myBodyIsHypermobileAndLovesToMove + alwaysHaveFun + 
    alwaysSerious + alwaysDepressed + alwaysLoveAndCareForYourself + 
    extremelyGoodAbilityToSense + alwaysHaveDigestiveProblems + 
    iAmIntenselyInterestedInOtherPeople + iRarelyWakeUpFeelingRested + 
    iFindMostThingsAmusing + iAmAlwaysCommittedAndInvolved + 
    iDoNotThinkThatTheWorldI

The fact that the rMSE of the stepwise model (7.33) is higher than that of the original full model (6.67) is somewhat unusual. Generally, stepwise regression aims to simplify the model by retaining significant predictors and removing redundant ones, potentially improving the model's performance. However, this higher rMSE indicates that the full model's complexity was necessary to capture important relationships within the data.

## Question 4 (2 Mark)
Although stepwise regression has reduced the model complexity significantly, the model still contains a lot of variables that we want to remove. Therefore, you are interested in lightweight linear regression models with ONLY TWO predictors. Write a script to automatically find the best lightweight model which corresponds to the model with the least **rMSE** on the training dataset. Compare the **rMSE** of the best lightweight model with the **rMSE** of the full model - ```lm.fit``` - that you built previously. Give an explanation for these results based on consideration of the predictors involved.

In [15]:
# ANSWER BLOCK
# Convert categorical variables to factors
for (col in colnames(train_df)) {
  if (is.character(train_df[[col]])) {
    train_df[[col]] <- as.factor(train_df[[col]])
  }
}

# Find the best lightweight model with only two predictors
best_rmse <- Inf
best_model <- NULL
best_predictors <- NULL

predictors <- colnames(train_df)[colnames(train_df) != "happiness"]

for (i in 1:(length(predictors) - 1)) {
  for (j in (i + 1):length(predictors)) {
    formula <- as.formula(paste("happiness ~", predictors[i], "+", predictors[j]))
    model <- lm(formula, data = train_df)
    predictions <- predict(model, newdata = train_df)
    rmse <- calculate_rmse(train_df$happiness, predictions)
    
    if (rmse < best_rmse) {
      best_rmse <- rmse
      best_model <- model
      best_predictors <- c(predictors[i], predictors[j])
    }
  }
}

# Print the results
print(paste("Best lightweight model predictors:", paste(best_predictors, collapse = " + ")))
print(paste("Root Mean Squared Error (rMSE) of Best Lightweight Model:", best_rmse))
print(paste("Root Mean Squared Error (rMSE) of Full Model:", rmse_full_model))


[1] "Best lightweight model predictors: income + alwaysStressed"
[1] "Root Mean Squared Error (rMSE) of Best Lightweight Model: 7.88541130213321"
[1] "Root Mean Squared Error (rMSE) of Full Model: 6.67255660795814"


### ANSWER (TEXT)
The results indicate that the best lightweight model, which includes only the predictors "income" and "alwaysStressed," has a higher root mean squared error (rMSE) of approximately 7.89 compared to the full model's rMSE of approximately 6.67.
The lightweight model with only two predictors, "income" and "alwaysStressed," is simpler but less accurate than the full model. This higher rMSE in the lightweight model indicates that other predictors in the full model are contributing significantly to the prediction of happiness. The full model's ability to incorporate more variables allows it to better capture the complexities and interactions within the data, resulting in more accurate predictions.

## Question 5 (Libraries are allowed) (40 Marks)
As a Data Scientist, one of the key tasks is to build models $\textbf{most appropriate/closest}$ to the truth; thus, modelling will not be limited to the aforementioned steps in this assignment. To simulate for a realistic modelling process, this question will be in the form of a [Kaggle competition](https://www.kaggle.com/t/ad8c96e412254c138cbec1d9d1c09734) among students to find out who has the best model.

Thus, you **will be graded** by the **rMSE** performance of your model, the better your model, the higher your score. Additionally, you need to describe/document your thought process in this model building process, this is akin to showing your working properly for the mathematic sections. If you don't clearly document the reasonings behind the model you use, we will have to make some deductions on your scores.

This is the [video tutorial](https://www.youtube.com/watch?v=rkXc25Uvyl4) on how to join any Kaggle competition. 

When you optimize your model's performance, you can use any supervised model that you know and feature selection might be a big help as well. [Check the non-exhaustive set of R functions relevant to this unit](https://learning.monash.edu/mod/resource/view.php?id=2017193) for ideas for different models to try.

$\textbf{Note}$ Please make sure that we can install the libraries that you use in this part, the code structure can be:

```install.packages("some package", repos='http://cran.us.r-project.org')```

```library("some package")```

Remember that if we cannot run your code, we will have to give you a deduction. Our suggestion is for you to use the standard ```R version 3.6.1```

You also need to name your final model ``fin.mod`` so we can run a check to find out your performance. A good test for your understanding would be to set the previous $\textbf{BIC model}$ to be the final model to check if your code works perfectly.

In [ ]:
# Function to check and install packages if not already installed
install_if_missing <- function(packages) {
  for (package in packages) {
    if (!require(package, character.only = TRUE)) {
      install.packages(package, repos = 'http://cran.us.r-project.org')
      library(package, character.only = TRUE)
    }
  }
}

# List of required packages
required_packages <- c("caret", "xgboost", "randomForest", "e1071")

# Install missing packages
install_if_missing(required_packages)

In [ ]:
# Load required libraries
library(caret)
library(xgboost)
library(randomForest)
library(e1071)

In [ ]:
# Load the data
train_df <- read.csv("regression_train.csv")

# Handle missing values by imputing with the median
preProcValues <- preProcess(train_df, method = c("medianImpute", "zv"))

# Ensure no missing values after preprocessing
train_df <- na.omit(train_df)

# Convert 'happiness' column to numeric if necessary
train_df$happiness <- as.numeric(train_df$happiness)

# Define the training control with 5-fold cross-validation
train_control <- trainControl(
  method = "cv", 
  number = 5,
  savePredictions = "final",
  allowParallel = TRUE
)

# Train the Random Forest model with the best parameters
set.seed(100)
rf_model <- train(
  happiness ~ ., 
  data = train_df, 
  method = "rf", 
  trControl = train_control, 
  tuneGrid = expand.grid(mtry = 12),
  ntree = 500  # Set ntree directly here
)

# Train the SVM model with the best parameters
set.seed(100)
svm_model <- train(
  happiness ~ ., 
  data = train_df, 
  method = "svmRadial", 
  trControl = train_control, 
  tuneGrid = expand.grid(C = 5, sigma = 0.005)
)

# Define the best parameters for xgbTree
best_params <- expand.grid(
  nrounds = 500,
  max_depth = 3,
  eta = 0.05,
  gamma = 0.5,
  colsample_bytree = 0.5,
  min_child_weight = 5,
  subsample = 0.9
)

# Train the XGBoost model with the best parameters
set.seed(100)
xgb_model <- train(
  happiness ~ ., 
  data = train_df, 
  method = "xgbTree", 
  trControl = train_control, 
  tuneGrid = best_params, 
  nthread = 1,
  verbosity = 1
)

# Combine predictions from the individual models
pred_train <- data.frame(
  xgb = predict(xgb_model, newdata = train_df),
  rf = predict(rf_model, newdata = train_df),
  svm = predict(svm_model, newdata = train_df),
  happiness = train_df$happiness
)

In [6]:
# Build your final model here, use additional coding blocks if you need to
set.seed(100)
fin.mod <- train(
  happiness ~ ., 
  data = pred_train, 
  method = "lm", 
  trControl = trainControl(method = "cv", number = 5)
)

In [7]:
# Load in the test data.
test <- read.csv("regression_test.csv")
# Handle missing values by imputing with the median
test <- predict(preProcValues, test)
# Ensure no missing values after preprocessing
test <- na.omit(test)
# Make predictions on the test set
pred_test <- data.frame(
  xgb = predict(xgb_model, newdata = test),
  rf = predict(rf_model, newdata = test),
  svm = predict(svm_model, newdata = test)
)
# If you are using any packages that perform the prediction differently, please change this line of code accordingly.
pred.label <- predict(fin.mod, newdata = pred_test)
# put these predicted labels in a csv file that you can use to commit to the Kaggle Leaderboard
write.csv(
  data.frame("RowIndex" = seq(1, length(pred.label)), "Prediction" = pred.label),  
  "RegressionPredictLabel.csv", 
  row.names = FALSE
)

In [ ]:
## PLEASE DO NOT ALTER THIS CODE BLOCK, YOU ARE REQUIRED TO HAVE THIS CODE BLOCK IN YOUR JUPYTER NOTEBOOK SUBMISSION
## Please skip (don't run) this if you are a student
## For teaching team use only

tryCatch(
    {
        source("../supplimentary.R")
    },
    error = function(e){
        source("supplimentary.R")
    }
)

truths <- tryCatch(
    {
        read.csv("../regression_test_label.csv")
    },
    error = function(e){
        read.csv("regression_test_label.csv")
    }
)


RMSE.fin <- rmse(pred.label, truths$x)
cat(paste("RMSE is", RMSE.fin))

# Part 2 Classification (50 Marks)

A few thousand people were questioned in a [life and wellbeing survey](https://www.get-happier.com/) to build a model to predict happiness of an individual, but this time we want to predict a categorical score for perfect mental health, rather than a continuous score. You need to build 5-class classification models to optimally predict the variable in the survey dataset called 'perfectMentalHealth' based on any, or all, of the other survey question responses. 

You have been provided with two datasets, ```classification_train.csv``` and ```classification_test.csv```. Using these datasets, you hope to build a model that can predict 'perfectMentalHealth' using the other variables. ```classification_train.csv``` comes with the ground-truth target label (i.e. 'perfectMentalHealth' happiness classes) whereas `classification_test.csv` comes with independent variables (input information) only.

On the order of around 70 survey questions have been converted into predictor variables that can be used to predict 'perfectMentalHealth'. We do not list all the predictor names here, but their names given in the data header can clearly be linked to the survey questions. E.g. the predictor variable 'iDontFeelParticularlyPleasedWithTheWayIAm' corresponds to the survey question 'I don’t feel particularly pleased with the way I am.'

This question will also be in the form of a [Kaggle competition](https://www.kaggle.com/t/968d3b346acb47779771f47785c39e62) among students to find out who has the best model.

In [ ]:
# Function to check and install packages if not already installed
install_if_missing <- function(packages) {
  for (package in packages) {
    if (!require(package, character.only = TRUE)) {
      install.packages(package, repos = 'http://cran.us.r-project.org')
      library(package, character.only = TRUE)
    }
  }
}

# List of required packages
required_packages <- c("readr", "dplyr", "xgboost", "caret")

# Install missing packages
install_if_missing(required_packages)

In [ ]:
# Load necessary libraries
library(readr)
library(dplyr)
library(xgboost)
library(caret)

In [ ]:
# Load in the train and test classification data.
train <- read.csv('classification_train.csv')
test <- read.csv('classification_test.csv')

# Function to encode ordinal and binary encode specific columns
encode_data <- function(data, columns_levels, binary_columns) {
  # Ordinal encoding for specified columns
  for (col in names(columns_levels)) {
    levels <- columns_levels[[col]]
    data[[col]] <- factor(data[[col]], levels = levels, ordered = TRUE)
    data[[col]] <- as.numeric(data[[col]])  # Convert factor to numeric
  }
  
  # Binary encoding for yes/no and gender columns
  for (col in binary_columns) {
    data[[col]] <- ifelse(data[[col]] == "Yes" | data[[col]] == "Female", 1, 0)
  }
  
  # Convert all columns to numeric, handling NAs
  data <- data.frame(lapply(data, function(x) {
    x <- as.numeric(as.character(x))
    x[is.na(x)] <- -999  # Replace NA values with a placeholder
    return(x)
  }))
  
  return(data)
}

# Update levels for each ordinal column to include all unique values
columns_levels <- list(
  income = c("0 - 10k", "10k - 15k", "15k - 20k", "20k - 50k", 
             "50k - 80k", "80k - 120k", "120k - 150k", "150k - 200k", "200k above"),
  whatIsYourHeightExpressItAsANumberInMetresM = c("140 - 150", "150 - 155", "155 - 160", "160 - 165", "165 - 170", 
                                                  "170 - 175", "175 - 180", "180 - 185", "185 - 190", "190 above"),
  howDoYouReconcileSpiritualBeliefsWithScientificOrRationalThinki = c("They are separate", "They complement each other", "They conflict"),
  howOftenDoYouFeelSociallyConnectedWithYourPeersAndFriends = c("Never", "Rarely", "Sometimes", "Often", "Always"),
  doYouHaveASupportSystemOfFriendsAndFamilyToTurnToWhenNeeded = c("No", "Yes, I have both", "Yes, I have family", "Yes, I have friends"),
  howOftenDoYouParticipateInSocialActivitiesIncludingClubsSportsV = c("No", "Rarely", "At least once a month", "At least once a week", "More than three times a week"),
  doYouFeelComfortableEngagingInConversationsWithPeopleFromDiffer = c("Never", "Rarely", "Sometimes", "Always")
)

# Define binary columns
binary_columns <- c("doYouFeelASenseOfPurposeAndMeaningInYourLife104", 
                    "doYouFeelASenseOfPurposeAndMeaningInYourLife105",
                    "gender")

# Encode the data
train_data_encoded <- encode_data(train_data, columns_levels, binary_columns)
test_data_encoded <- encode_data(test_data, columns_levels, binary_columns)

# Select top features after performing Recursive Feature Elimination
top_features <- c("alwaysLoveAndCareForYourself", "lifeIsGood", 
                  "alwaysEngageInPreparingAndUsingYourSkillsAndTalentsInOrderToGai", 
                  "alwaysCalm", "alwaysStressed")

# Subset the data to include only the top features
train_data_top_features <- train_data_encoded %>% select(all_of(top_features))
test_data_top_features <- test_data_encoded %>% select(all_of(top_features))

# Convert original labels to a 0-based index
unique_labels <- sort(unique(train_data$perfectMentalHealth))
label_mapping <- setNames(seq_along(unique_labels) - 1, unique_labels)
train_labels <- as.numeric(label_mapping[as.character(train_data$perfectMentalHealth)])
test_labels <- as.numeric(label_mapping[as.character(test_data$perfectMentalHealth)])
num_classes <- length(unique_labels)

# Define the best parameters for XGBoost after model tuning
best_params <- list(
  objective = "multi:softmax",
  eval_metric = "merror",
  num_class = num_classes,
  max_depth = 5,
  min_child_weight = 0.8,
  gamma = 0.1,
  subsample = 0.7,
  colsample_bytree = 0.6,
  eta = 0.01
)

set.seed(123)

# Prepare data for XGBoost
dtrain <- xgb.DMatrix(data = as.matrix(train_data_top_features), label = train_labels)
dtest <- xgb.DMatrix(data = as.matrix(test_data_top_features))

In [ ]:
# Build your final model here, use additional coding blocks if you need to
fin.mod <- xgb.train(params = best_params, data = dtrain, nrounds = 1000)

# If you are using any packages that perform the prediction differently, please change this line of code accordingly.
pred_indices <- predict(fin.mod, dtest)
pred.label <- unique_labels[pred_indices + 1]  # Map back to original labels

# put these predicted labels in a csv file that you can use to commit to the Kaggle Leaderboard
write.csv(
    data.frame("RowIndex" = seq(1, length(pred.label)), "Prediction" = pred.label),  
    "ClassificationPredictLabel.csv", 
    row.names = F
)

In [ ]:
## PLEASE DO NOT ALTER THIS CODE BLOCK, YOU ARE REQUIRED TO HAVE THIS CODE BLOCK IN YOUR JUPYTER NOTEBOOK SUBMISSION
## Please skip (don't run) this if you are a student
## For teaching team use only

truths <- tryCatch(
    {
        read.csv("../classification_test_label.csv")
    },
    error = function(e){
        read.csv("classification_test_label.csv")
    }
)

f1_score <- F1_Score(truths$x, pred.label)
cat(paste("f1_score is", f1_score))